# 1. Data Preparation

In [ ]:
from tensorflow import keras

# Load dataset
(x_train, y_train), (x_test, y_test) = keras.datasets.mnist.load_data()

# Flatten images (28x28 → 784)
x_train_flat = x_train.reshape(-1, 28 * 28).astype('float32') / 255.0
x_test_flat = x_test.reshape(-1, 28 * 28).astype('float32') / 255.0

print("Training shape:", x_train_flat.shape)
print("Test shape:", x_test_flat.shape)

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
Training shape: (60000, 784)
Test shape: (10000, 784)


# 2. Experiment with different Activation Functions

# 2.1 Relu Activation Function

In [ ]:
from tensorflow import keras
from tensorflow.keras import layers

model1 = keras.Sequential([
    layers.Dense(64, activation='relu', input_shape=(784,)),   # hidden layer 1
    layers.Dense(32, activation='relu'),                        # hidden layer 2
    layers.Dense(10, activation='softmax')                      # output (10 classes)
])



model1.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

# model1.summary()

# Train the model
history1 = model1.fit(
    x_train_flat, y_train,
    epochs=5,
    batch_size=32,
    validation_split=0.2
)

# Evaluate the model
loss1, acc1 = model1.evaluate(x_test_flat, y_test)

print("Model 1 Test Accuracy:", acc1)

Epoch 1/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 8s 5ms/step - accuracy: 0.9071 - loss: 0.3183 - val_accuracy: 0.9452 - val_loss: 0.1882
Epoch 2/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step - accuracy: 0.9562 - loss: 0.1483 - val_accuracy: 0.9613 - val_loss: 0.1344
Epoch 3/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9677 - loss: 0.1081 - val_accuracy: 0.9629 - val_loss: 0.1220
Epoch 4/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step - accuracy: 0.9734 - loss: 0.0867 - val_accuracy: 0.9682 - val_loss: 0.1099
Epoch 5/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9780 - loss: 0.0714 - val_accuracy: 0.9697 - val_loss: 0.1054
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9714 - loss: 0.0950
Model 1 Test Accuracy: 0.9714000225067139


Test Accuracy: 0.9714
- Observations:
  - The model performs very well on MNIST (~97% accuracy)
  - ReLU activation helps fast convergence

# 2.2 Tanh Activation Function

In [ ]:
model2 = keras.Sequential([
    layers.Dense(64, activation='tanh', input_shape=(784,)),
    layers.Dense(32, activation='tanh'),
    layers.Dense(10, activation='softmax')
])

model2.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

# Train
history2 = model2.fit(
    x_train_flat, y_train,
    epochs=5,
    batch_size=32,
    validation_split=0.2
)

# Evaluate
loss2, acc2 = model2.evaluate(x_test_flat, y_test)

print("Model 2 Test Accuracy:", acc2)

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - accuracy: 0.9040 - loss: 0.3414 - val_accuracy: 0.9453 - val_loss: 0.1893
Epoch 2/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step - accuracy: 0.9538 - loss: 0.1537 - val_accuracy: 0.9562 - val_loss: 0.1419
Epoch 3/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 10s 4ms/step - accuracy: 0.9680 - loss: 0.1089 - val_accuracy: 0.9621 - val_loss: 0.1222
Epoch 4/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 7s 5ms/step - accuracy: 0.9745 - loss: 0.0835 - val_accuracy: 0.9664 - val_loss: 0.1116
Epoch 5/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step - accuracy: 0.9793 - loss: 0.0676 - val_accuracy: 0.9675 - val_loss: 0.1071
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9682 - loss: 0.1049
Model 2 Test Accuracy: 0.9682000279426575


Test Accuracy: 0.9682
- Observations:

    - Accuracy is slightly lower than the when we used ReLU activation function (0.9714 → 0.9682)

    - Training is still stable, but learning is a bit less efficient than ReLU

# 3. Experiment with Learning Rate

we will keep the same model but we will change the optimizer

In [ ]:
from tensorflow.keras.optimizers import Adam

model3 = keras.Sequential([
    layers.Dense(64, activation='relu', input_shape=(784,)),
    layers.Dense(32, activation='relu'),
    layers.Dense(10, activation='softmax')
])

optimizer = Adam(learning_rate=0.0001)
model3.compile(
    optimizer=optimizer,
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

history3 = model3.fit(
    x_train_flat, y_train,
    epochs=5,
    batch_size=32,
    validation_split=0.2
)

loss3, acc3 = model3.evaluate(x_test_flat, y_test)

print("Model 3 Test Accuracy:", acc3)

Epoch 1/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 8s 5ms/step - accuracy: 0.7771 - loss: 0.8459 - val_accuracy: 0.9026 - val_loss: 0.3707
Epoch 2/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.9051 - loss: 0.3435 - val_accuracy: 0.9223 - val_loss: 0.2861
Epoch 3/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step - accuracy: 0.9228 - loss: 0.2782 - val_accuracy: 0.9311 - val_loss: 0.2469
Epoch 4/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 10s 4ms/step - accuracy: 0.9322 - loss: 0.2417 - val_accuracy: 0.9396 - val_loss: 0.2205
Epoch 5/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 7s 5ms/step - accuracy: 0.9391 - loss: 0.2163 - val_accuracy: 0.9448 - val_loss: 0.2042
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9428 - loss: 0.2045
Model 3 Test Accuracy: 0.942799985408783


Test Accuracy: 0.9438
- Observations:
    - This model uses a smaller learning rate (0.0001)
    - Training is more stable but significantly slower to converge
    - Accuracy is lower than previous models (0.9714 → 0.9438)
    - The model learns more cautiously but underperforms in limited epochs

# 4. Final Comparison of All Models

| Model | Activation | Learning Rate | Test Accuracy |
|------|------------|---------------|---------------|
| Model 1 | ReLU | 0.001 (default) | 0.9714 |
| Model 2 | Tanh | 0.001 (default) | 0.9682 |
| Model 3 | ReLU | 0.0001 | 0.9438 |

# Conclusion:
- ReLU performs better than Tanh in this dataset
- Default learning rate (0.001) is the most effective for this problem
- Reducing learning rate too much slows learning and reduces performance within limited epochs